### ISSUE TITLE
debounce addon throttles instead of debounces

### ISSUE THREAD

**Issue by NervosaX** [cerebral/cerebral-addons#19]
**What I expect debounce to do:**

Debounce should halt firing the complete signal path until the signal has not been called for x milliseconds. As an example, if I have a debounce on an input, I would want to fire the signal on every character, but only have the complete path followed as soon as I have stopped typing for 300ms or so.

Eg:

```
debounce(300, [
	[saveContact, {
		success: [],
		error: []
	}]
], { immediate: false })
```

**What debounce is doing:**

The debounce addon is - with the above example - calling saveContact every 300ms until I stop typing. This is more in line with a throttle function.

** Note: **

An immediate call on a debounce also feels very weird. I can't think of terribly many usecases where you would want that. An immediate call on a throttle would make more sense however.

**Comment by garth** [comment]
Did you look at the unit tests: https://github.com/cerebral/cerebral-addons/blob/master/test/factories/debounce.js As far as I can tell it should be working as expected. I have it running in production apps where I want to debounce key strokes.

There was quite a discussion about immediate firing a couple of months back. The logic is that for a one time event why wait, but there's always the `{ immediate: false }` option when you need it.

**Comment by garth** [comment]
The different between throttle and debounce, is that debounce will drop signals between the 300ms whilst throttle will queue them and eventually run them all.

**Comment by NervosaX** [comment]
Hrm, definitely looks like the test works how I'd expect debounce to work. I'll try and set up a smaller test case and see if I can replicate the issue.

That's a reasonable reason.

**Comment by NervosaX** [comment]
```javascript
	debounce(1000, [
		function() {
			console.log("foo");
		}
	])

```

Just running this, it seems to keep calling "foo" over and over if I mash the click button to hit the signal over and over.

**Comment by garth** [comment]
Maybe it's related to the unit tests not using real signals.... 

If you have time, can you try to make a breaking unit test?

If it helps the operators tests show how to test with a real controller/signal setup (https://github.com/cerebral/cerebral-addons/blob/master/test/operators.js)

**Comment by NervosaX** [comment]
Sure thing, I'll see how I go.

**Comment by NervosaX** [comment]
Initial attempts at a naive real-signal attempt seem to not break. 

```javascript
function increaseCount ({ state }) {
  state.set('count', state.get('count') + 1)
}

controller.addSignals({
  increase: { chain: [
    debounce(1, [
      increaseCount
    ], { immediate: false })
  ], sync: true }
})

const signals = controller.getSignals()
let tree

describe('debounce()', function () {
  beforeEach(function () {
    tree = controller.model.tree
    tree.set({
      count: 0
    })
    tree.commit()
  })

  it('should not call increase more than twice when immediate', function (done) {
    signals.increase()
    signals.increase()
    signals.increase()
    signals.increase()
    signals.increase()
    signals.increase()
    signals.increase()
    signals.increase()

    setTimeout(function () {
      console.log(tree.get('count'))
      done()
    }, 10)
  })
```

The count is returning 2 for immediate = true, and 1 for immediate = false.

This suggests the code is perfectly fine. I've tried making sync false, with no changes. My addons is up to date, so I just have to keep digging deeper...

**Comment by NervosaX** [comment]
@garth 
Okay, I think I've replicated the issue in a small test app.

https://github.com/nervosax/cerebral-issue/tree/bug/debounce

(Make sure you're on the bug/debounce branch).

I have set up a debounce action to be update a count every 2 seconds. In my layout component there is a state toggle for allowing the interval to run every 1 second. I would expect that once I've clicked the "start" button on this page, the count would not go up until I clicked stop. Instead the count (erratically I think) continues to go up. 

![image](https://cloud.githubusercontent.com/assets/1562807/13685411/7164e0a2-e764-11e5-8d24-587410758793.png)
(The image shows running it once and clicking stop once).
